thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js

```diff
@@ -276,11 +292,65 @@ export default Ember.Controller.extend({
     };
   },
 
+  _setPropertiesUndo(properties, action, message) {
+    const beforeImage = this.getProperties(...Object.keys(properties));
+    const { undoLog } = this.getProperties('undoLog');
+
+    const timestamp = moment().valueOf();
+
+    // merge checkpoint if last action is the same within a time window
+    if (!_.isEmpty(undoLog)) {
+      const lastLog = undoLog[undoLog.length - 1];
+      if (lastLog.action === action &&
+          lastLog.message === message &&
+          lastLog.timestamp >= timestamp - ROOTCAUSE_UNDO_MERGE_TIMEOUT) {
+
+        const newLogEntry = Object.assign({}, lastLog, { timestamp });
+        const newUndoLog = [..._.slice(undoLog, 0, undoLog.length - 1), newLogEntry];
+
+        this.setProperties(Object.assign({}, properties, { undoLog: newUndoLog }));
+        return;
+      }
+    }
+
+    const logEntry = {
+      beforeImage,
+      timestamp,
+      action,
+      message
+    };
+
+    const newUndoLog = [...undoLog, logEntry];
+
+    if (!_.isEqual(beforeImage, properties)) {
+      this.setProperties(Object.assign({}, properties, { undoLog: newUndoLog }));
+    }
+  },
+
+  _undoLast() {
+    const { undoLog } = this.getProperties('undoLog');
+
+    if (_.isEmpty(undoLog)) { return; }
+
+    const { beforeImage } = undoLog[undoLog.length - 1];
+
+    const newUndoLog = _.slice(undoLog, 0, undoLog.length - 1);
+
+    this.setProperties(Object.assign({}, beforeImage, { undoLog: newUndoLog }));
+  },
+
   //
   // Actions
   //
 
   actions: {
+    /**
+     * Undoes the last tracked user action.
+     */
+    onUndo() {
+      this._undoLast();
```

Comment:
Since this method is only used here, I would suggest moving the logic to here and remove the method definition above.